### PR TITLE
Reject invalid FFDHE and ECDHE key shares with SSL_AD_ILLEGAL_PARAMETER alert (RFC 8446)

### PR DESCRIPTION
**Pull request by ldlb9527**
Fixes: #25402 

The test results after modification are as follows:

<details>
<summary>dhe</summary>

```sh
$ PYTHONPATH=. python scripts/test-dhe-rsa-key-exchange-with-bad-messages.py
sanity ...
OK

invalid dh_Yc value - 1 ...
OK

padded Client Key Exchange ...
OK

invalid dh_Yc value - 8192b ...
OK

invalid dh_Yc value - p ...
OK

invalid dh_Yc value - p-1 ...
OK

truncated dh_Yc value ...
OK

invalid dh_Yc value - 0 ...
OK

invalid dh_Yc value - missing ...
OK

sanity ...
OK

Check if server properly verifies received Client Key Exchange
message. That the extra data (pad) at the end is noticed, that
too short message is rejected and a message with "obviously"
wrong client key share is rejected
Test end
====================
version: 6
====================
TOTAL: 10
SKIP: 0
PASS: 10
XFAIL: 0
FAIL: 0
XPASS: 0
====================
```

</details>

<details>
<summary>ecdhe</summary>

```sh
$ PYTHONPATH=. python scripts/test-ecdhe-rsa-key-exchange-with-bad-messages.py
sanity ...
OK

invalid point (self-inconsistent, y all zero) ...
OK

invalid point (0, 0) ...
OK

invalid point (self-inconsistent) ...
OK

truncated ecdh_Yc value ...
OK

padded Client Key Exchange ...
OK

invalid point (0, 1) ...
OK

sanity ...
OK

Test end
====================
version: 5
====================
TOTAL: 8
SKIP: 0
PASS: 8
XFAIL: 0
FAIL: 0
XPASS: 0
====================
```

</details>

### CODE REVIEW

File: ssl/statem/statem_srvr.c [hunk 1]
```diff
@@ -3084,7 +3084,7 @@ static int tls_process_cke_dhe(SSL_CONNECTION *s, PACKET *pkt)
     }
 
     if (!EVP_PKEY_set1_encoded_public_key(ckey, data, i)) {
-        SSLfatal(s, SSL_AD_INTERNAL_ERROR, ERR_R_INTERNAL_ERROR);
+        SSLfatal(s, SSL_AD_ILLEGAL_PARAMETER, ERR_R_DH_LIB);
```
Comment (t8m) [inline ssl/statem/statem_srvr.c:3087]: Please use SSL_R_BAD_KEY_SHARE instead of ERR_R_DH_LIB and same in the ERR_R_EC_LIB case below.

**Review by t8m — CHANGES_REQUESTED** [review]

**Comment by t8m** [comment]
Would it be hard to add a test case?

File: ssl/statem/statem_srvr.c [hunk 2]
```diff
@@ -3138,7 +3138,7 @@ static int tls_process_cke_ecdhe(SSL_CONNECTION *s, PACKET *pkt)
         }
 
         if (EVP_PKEY_set1_encoded_public_key(ckey, data, i) <= 0) {
-            SSLfatal(s, SSL_AD_INTERNAL_ERROR, ERR_R_EC_LIB);
+            SSLfatal(s, SSL_AD_ILLEGAL_PARAMETER, ERR_R_EC_LIB);
```
Comment (vdukhovni) [inline ssl/statem/statem_srvr.c:3141]: This PR brings to light a more interesting potential issue than that which alert is raised.

* In the DHE case the function `EVP_PKEY_set1_encoded_public_key` is expected to return a boolean value (0 or non-zero)
* In the ECDHE case the **same** function is expected to return a positive value for success and a zero or negative value for failure.

One convention used in some places in the code is that 0 is expected failure, while negative is unexpected failure, and they can be treated differently, allowing, say, malloc failures to still report internal error, while actual problems with the input might report the illegal parameter alert.

The manpage says:
```
RETURN VALUES
    EVP_PKEY_set1_encoded_public_key() returns 1 for success and 0 or a
    negative value for failure.
```
So the DHE case is not strictly correct, but the implementation always returns 0 or 1 for legacy keys: https://github.com/openssl/openssl/blob/0958f5a5bc47e2cca907a3bfaf14059c020324fd/crypto/evp/p_lib.c#L1386-L1404
Leaving just provided keys as a potential source of negative return values, but that ultimately takes us to: https://github.com/openssl/openssl/blob/0958f5a5bc47e2cca907a3bfaf14059c020324fd/crypto/evp/keymgmt_meth.c#L485 so up to whatever the provider does. This time documentation states: https://github.com/openssl/openssl/blob/0958f5a5bc47e2cca907a3bfaf14059c020324fd/doc/man7/provider-keymgmt.pod?plain=1#L478-L479

So, bottom line, the negative values are historical, and at this point probably can't be brought back without breaking the API.
So, in the end, no way to distinguish input errors from resource errors. Oh, well...

@t8m should we just fix the nits with the 3rd argument and merge without waiting for the author?

Comment (t8m) [reply ssl/statem/statem_srvr.c:3141]: > @t8m should we just fix the nits with the 3rd argument and merge without waiting for the author?

If you want to, you can open a new PR. Or push a fixup commit to this author's branch.

Comment (vdukhovni) [reply ssl/statem/statem_srvr.c]: Done.

**Review by vdukhovni — APPROVED** [review]

**Comment by vdukhovni** [comment]
Fixed the error number and rebased to latest master.

**Review by t8m — APPROVED** [review]

**Comment by t8m** [comment]
Merged to all the active branches. Thank you for your contribution.
